platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -434,41 +435,59 @@ public void endWriting() throws IOException {
                     oneTransactionsData.keyValuesView().toList();
             final int size = bucketUpdates.size();
             final Queue<ReadBucketResult<K>> queue = new ConcurrentLinkedQueue<>();
+            final AtomicBoolean endWritingRunning = new AtomicBoolean(true);
             for (final IntObjectPair<BucketMutation<K>> keyValue : bucketUpdates) {
                 final int bucketIndex = keyValue.getOne();
                 final BucketMutation<K> bucketMap = keyValue.getTwo();
-                flushExecutor.execute(() -> readUpdateQueueBucket(bucketIndex, bucketMap, queue));
+                flushExecutor.execute(() -> {
+                    if (endWritingRunning.get()) {
+                        readUpdateQueueBucket(bucketIndex, bucketMap, queue);
+                    }
+                });
             }
-            //  write to files
-            fileCollection.startWriting();
-            int processed = 0;
-            while (processed < size) {
-                final ReadBucketResult<K> res = queue.poll();
-                if (res == null) {
-                    continue;
-                }
-                if (res.error != null) {
-                    throw new RuntimeException(res.error);
+            try {
+                //  write to files
+                fileCollection.startWriting();
+                int processed = 0;
+                while (processed < size) {
+                    final ReadBucketResult<K> res = queue.poll();
+                    if (res == null) {
+                        continue;
```

Comment:
This could result in a very expensive spin-wait when the queue is empty for more than a few microseconds.
Might be good to call `Thread.sleep(0, 100)` (a 100ns sleep) or at least `Thread.yield()` before `continue` so we don't get priority amplification issues.

Reply:
`Thread.onSpinWait()` is already used in `ReusableBucketPool.releaseBucket()`. Could be used here as well.

Reply:
Added